### PR TITLE
Get visit service

### PR DESCRIPTION
**Pull request by killianherbunot**

### CODE REVIEW

File: src/Client/ApiClient.php [hunk 1]
```diff
@@ -9,18 +9,18 @@
...
-     * @param mixed  $resourceData
      *
      * @return string
      */
-    public function post($resourceName, $resourceData);
+    public function get($resourceName);
 
     /**
      * @param string $resourceName
+     * @param object $resourceData
```
Comment (arnaud-23) [inline src/Client/ApiClient.php]: array

File: src/Gateways/VisitGateway.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+use OpenClassrooms\FrontDesk\Models\Visit;
+
+/**
+ * @author Killian Herbunot <killian.herbunot@openclassrooms.com>
+ */
+interface VisitGateway
+{
+    /**
+     * @param integer $personId
```
Comment (arnaud-23) [inline src/Gateways/VisitGateway.php]: what about "from" and "to" ?

File: src/Models/Impl/VisitBuilderImpl.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ */
+class VisitBuilderImpl implements VisitBuilder
+{
+    /**
+     * @var VisitImpl
+     */
+    private $visit;
+
+    /**
+     * @return VisitBuilder
```
Comment (arnaud-23) [inline src/Models/Impl/VisitBuilderImpl.php]: inheritdoc (same for the entire method of this class)

File: src/Repository/VisitRepository.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+use OpenClassrooms\FrontDesk\Models\Impl\VisitBuilderImpl;
+use OpenClassrooms\FrontDesk\Models\VisitBuilder;
+use OpenClassrooms\FrontDesk\Services\Impl\InvalidTotalCountException;
+
+/**
+ * @author Killian Herbunot <killian.herbunot@openclassrooms.com>
+ */
+class VisitRepository implements VisitGateway
+{
+    const RESOURCE_NAME = 'desk/people/';
```
Comment (arnaud-23) [inline src/Repository/VisitRepository.php]: we don't add the last "/" for other resource name do the same here

File: tests/Repository/VisitRepositoryTest.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+use OpenClassrooms\FrontDesk\Doubles\Client\Impl\ApiClientMock;
+use OpenClassrooms\FrontDesk\Doubles\Models\PersonStub1;
+use OpenClassrooms\FrontDesk\Doubles\Models\VisitStub1;
+
+/**
+ * @author Killian Herbunot <killian.herbunot@openclassrooms.com>
+ */
+class VisitRepositoryTest extends \PHPUnit_Framework_TestCase
+{
+    const FROM = "2016-08-12 15:30:06";
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: single quote

File: tests/Repository/VisitRepositoryTest.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    const TO = "2016-09-12 15:30:06";
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: same

File: tests/Repository/VisitRepositoryTest.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    private $visitRepository;
+
+    /**
+     * @test
+     *
+     * @expectedException \OpenClassrooms\FrontDesk\Services\Impl\InvalidTotalCountException
+     */
+    public function checkTotalCount()
+    {
+        ApiClientMock::$response = json_encode(["visits" => [new VisitStub1()], "total_count" => 0]);
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: same

File: tests/Repository/VisitRepositoryTest.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+        $this->visitRepository->findAllByPersonId(PersonStub1::ID);
+    }
+
+    /**
+     * @test
+     */
+    public function findAllByPersonIdWithoutParameters_ReturnVisits()
+    {
+        ApiClientMock::$response = json_encode(["visits" => [new VisitStub1()], "total_count" => 1]);
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: same

File: tests/Repository/VisitRepositoryTest.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        $this->assertEquals(json_encode([new VisitStub1()]), json_encode($visits));
+    }
+
+    /**
+     * @test
+     */
+    public function findAllByPersonIdWithParameters_ReturnVisits()
+    {
+        list($from, $to) = $this->initFromAndTo();
+        ApiClientMock::$response = json_encode(["visits" => [new VisitStub1()], "total_count" => 1]);
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: and same

File: tests/Repository/VisitRepositoryTest.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * @test
+     */
+    public function findAllByPersonIdWithoutParameters_ReturnVisits()
+    {
+        ApiClientMock::$response = json_encode(["visits" => [new VisitStub1()], "total_count" => 1]);
+
+        $visits = $this->visitRepository->findAllByPersonId(PersonStub1::ID);
+        $this->assertEquals(json_encode([new VisitStub1()]), json_encode($visits));
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: what about the use of the assertVisit in VisitTestCase trait ?

File: tests/Services/Impl/VisitServiceImplTest.php [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    public function getVisitsWithoutParameters_ReturnVisits()
+    {
+        $visits = $this->service->getVisits(PersonStub1::ID);
+        $this->doTestVisit($visits);
+    }
+
+    /**
+     * @param Visit[] $visits
+     */
+    public function doTestVisit($visits)
```
Comment (arnaud-23) [inline tests/Services/Impl/VisitServiceImplTest.php]: assertVisit**s**

File: tests/Services/Impl/VisitServiceImplTest.php [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        $this->assertEquals(0, count($visits) - count(VisitGatewayMock::$visits));
```
Comment (arnaud-23) [inline tests/Services/Impl/VisitServiceImplTest.php]: ?

File: tests/Services/Impl/VisitServiceImplTest.php (hunk 11 above)
Comment (killianherbunot) [inline tests/Services/Impl/VisitServiceImplTest.php]: not sure, i'm only test one visit at a time.

File: src/Gateways/VisitGateway.php (hunk 2 above)
Comment (killianherbunot) [inline src/Gateways/VisitGateway.php]: done

File: tests/Repository/PlanRepositoryTest.php [hunk 13]
```diff
@@ -36,12 +40,19 @@ public function findAllByPersonId_ReturnPlans()
         ApiClientMock::$response = json_encode(["plans" => [new PlanStub1()], "total_count" => 1]);
 
         $plans = $this->planRepository->findAllByPersonId(PersonStub1::ID);
-        $this->assertEquals(json_encode([new PlanStub1()]), json_encode($plans));
+        $this->assertEquals(0, count($plans) - count(PlanGatewayMock::$plans));
```
Comment (arnaud-23) [inline tests/Repository/PlanRepositoryTest.php]: useless

File: tests/Repository/PlanRepositoryTest.php [hunk 14]
```diff
@@ -36,12 +40,19 @@ public function findAllByPersonId_ReturnPlans()
...
+            $planStub = '\OpenClassrooms\FrontDesk\Doubles\Models\PlanStub'.++$i;
+            $this->assertPlan(new $planStub(), $plan);
+        }
     }
 
     public function setUp()
     {
         $this->planRepository = new PlanRepository();
         $this->planRepository->setApiClient(new ApiClientMock());
+        PlanGatewayMock::$plans = new PlanStub1();
```
Comment (arnaud-23) [inline tests/Repository/PlanRepositoryTest.php]: same

File: tests/Repository/VisitRepositoryTest.php [hunk 15]
```diff
@@ -38,10 +43,24 @@ public function checkTotalCount()
...
-        $this->assertEquals(json_encode([new VisitStub1()]), json_encode($visits));
+        $this->doTestVisits($visits);
+    }
+
+    /**
+     * @param Visit[] $visits
+     */
+    public function doTestVisits($visits)
+    {
+        $this->assertEquals(0, count($visits) - count(VisitGatewayMock::$visits));
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: useless

File: tests/Repository/VisitRepositoryTest.php [hunk 16]
```diff
@@ -74,5 +93,6 @@ public function setUp()
     {
         $this->visitRepository = new VisitRepository();
         $this->visitRepository->setApiClient(new ApiClientMock());
+        VisitGatewayMock::$visits = new VisitStub1();
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: useless

File: tests/Repository/VisitRepositoryTest.php [hunk 17]
```diff
@@ -38,10 +43,24 @@ public function checkTotalCount()
...
 
         $visits = $this->visitRepository->findAllByPersonId(PersonStub1::ID);
-        $this->assertEquals(json_encode([new VisitStub1()]), json_encode($visits));
+        $this->doTestVisits($visits);
+    }
+
+    /**
+     * @param Visit[] $visits
+     */
+    public function doTestVisits($visits)
```
Comment (arnaud-23) [inline tests/Repository/VisitRepositoryTest.php]: assertVisit**s** and you can move it on VisitTestCase Trait

File: tests/Doubles/Models/VisitTestCase.php [hunk 18]
```diff
@@ -10,6 +10,18 @@
  */
 trait VisitTestCase
 {
+    /**
+     * @param Visit[] $visits
+     */
+    public function assertVisits($visits)
```
Comment (arnaud-23) [inline tests/Doubles/Models/VisitTestCase.php]: (array $visits) :)

Comment (arnaud-23) [inline tests/Doubles/Models/VisitTestCase.php]: you need it in your VisitServiceImpl too

**Pull request merged**